### PR TITLE
Finished constraints #LMR-35

### PR DESCRIPTION
**Pull request by marvec**
It is not top notch, there are non-trivial code duplicates, but given the limited time I planned to spent here, I believe we are good to go.

### CODE REVIEW

File: api/src/main/java/io/lumeer/engine/api/constraint/ListConstraintType.java [hunk 1]
```diff
@@ -38,9 +43,54 @@
...
    }
 
+   private BiFunction<Object, Class, Object> getEncodeFunction(final boolean oneOf) {
+      return (o, t) -> {
+         if (oneOf) {
+            if (t == null || t == String.class) {
+               return o.toString();
+            }
+         } else {
+            if (t == null || String[].class.isAssignableFrom(t)) {
```
Comment (livthomas) [inline api/src/main/java/io/lumeer/engine/api/constraint/ListConstraintType.java]: Is there always going to be an array of strings?

File: api/src/main/java/io/lumeer/engine/api/constraint/NumberConstraintType.java [hunk 2]
```diff
@@ -51,88 +54,130 @@
...
    private NumberFormat numberFormat = NumberFormat.getNumberInstance();
-   private NumberFormat integerNumberFormat = NumberFormat.getNumberInstance();
+   private NumberFormat bigNumberFormat = NumberFormat.getNumberInstance();
+   private NumberFormat integerNumberFormat = NumberFormat.getIntegerInstance();
+   private NumberFormat currencyNumberFormat = NumberFormat.getNumberInstance();
 
    public NumberConstraintType() {
-      integerNumberFormat.setParseIntegerOnly(true);
+      ((DecimalFormat) currencyNumberFormat).setParseBigDecimal(true);
+      ((DecimalFormat) bigNumberFormat).setParseBigDecimal(true);
```
Comment (livthomas) [inline api/src/main/java/io/lumeer/engine/api/constraint/NumberConstraintType.java]: What is the difference between `bigNumberFormat` and `currencyNumberFormat`?

Comment (marvec) [inline api/src/main/java/io/lumeer/engine/api/constraint/NumberConstraintType.java]: For now there is none, in the future there might be. I am not sure if we want to add currency or so. So I kept that separated.

File: api/src/main/java/io/lumeer/engine/api/constraint/ListConstraintType.java (hunk 1 above)
Comment (marvec) [inline api/src/main/java/io/lumeer/engine/api/constraint/ListConstraintType.java]: For tags type it must be.

File: api/src/main/java/io/lumeer/engine/api/constraint/CaseConstraintType.java [hunk 3]
```diff
@@ -47,6 +52,30 @@
       return Collections.singleton(CONSTRAINT_PREFIX);
    }
 
+   static BiFunction<Object, Class, Object> getEncodeFunction() {
+      return (o, t) -> {
+         if (t != null && t != String.class) {
+            return null;
```
Comment (marvec) [inline api/src/main/java/io/lumeer/engine/api/constraint/CaseConstraintType.java]: We accept only String.class request. So either the type is not specified or String is requested.

File: api/pom.xml [hunk 4]
```diff
@@ -97,10 +97,33 @@
...
          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <configuration>
+               <forkCount>1C</forkCount>
+               <reuseForks>true</reuseForks>
+               <redirectTestOutputToFile>false</redirectTestOutputToFile>
+               <systemPropertyVariables>
+                  <log4j.skipJansi>true</log4j.skipJansi>
+               </systemPropertyVariables>
```
Comment (marvec) [inline api/pom.xml]: I just copy-pasted that from war module.

File: api/src/main/java/io/lumeer/engine/api/constraint/CaseConstraintType.java (hunk 3 above)
Comment (livthomas) [inline api/src/main/java/io/lumeer/engine/api/constraint/CaseConstraintType.java]: Oh, I see it now. It was just confusing at first sight.

File: api/src/main/java/io/lumeer/engine/api/constraint/DateTimeConstraintType.java [hunk 5]
```diff
@@ -48,6 +51,38 @@
       return REGISTERED;
    }
 
+   private BiFunction<Object, Class, Object> getEncodeFunction(final DateFormat format) {
```
Comment (marvec) [inline api/src/main/java/io/lumeer/engine/api/constraint/DateTimeConstraintType.java]: I did think about it. I even considered putting it to a separate class for all constraints. But something prevented me to do so. Will have a look...

File: api/pom.xml (hunk 4 above)
Comment (marvec) [inline api/pom.xml]: I think that plugin configuration is not inherited from the parent pom. Moreover, I changed the values to run tests in multiple threads at the same time.

**Pull request merged**